libbpf-tools/mountsnoop.bpf.c

```diff
@@ -0,0 +1,131 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2021 Hengqi Chen */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "mountsnoop.h"
+
+#define MAX_ENTRIES 10240
+
+const volatile pid_t target_pid = 0;
+static struct event empty = {};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct arg);
+} args SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u32);
+	__type(value, struct event);
+} values SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(__u32));
+	__uint(value_size, sizeof(__u32));
+} events SEC(".maps");
+
+static int probe_entry(const char *src, const char *dest, const char *fs,
+		       __u64 flags, const char *data, enum op op)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = (__u32)pid_tgid;
+	struct arg arg = {};
+
+	if (target_pid && target_pid != pid)
+		return 0;
+
+	arg.ts = bpf_ktime_get_ns();
+	arg.flags = flags;
+	arg.src = src;
+	arg.dest = dest;
+	arg.fs = fs;
+	arg.data= data;
+	arg.op = op;
+	bpf_map_update_elem(&args, &tid, &arg, BPF_ANY);
+	bpf_map_update_elem(&values, &tid, &empty, BPF_ANY);
```

Comment:
hm... I don't think you need two separate maps for this. Just store both args and event in one struct, but at the end submit only the event part. Makes everything simpler and more efficient.

Reply:
hm... thinking some more, you don't even need hashmap for event/values.
1. Either use per-cpu array for temporary storage (you need event only during probe_exit execution, so using per-cpu is fine, because that BPF program is not going to be migrated to another CPU nor will it be interrupted by the same tracepoint).
2. Or use BPF ringbuf and just avoid extra map for struct event. Do bpf_ringbuf_reserve(), fill out data, submit. Done.

Reply:
Thanks. Will look into this later.

`bpf_map_update_elem(&values, &tid, &empty, BPF_ANY);` this one is used to avoid stack limit of BPF programs.

Reply:
> `bpf_map_update_elem(&values, &tid, &empty, BPF_ANY);` this one is used to avoid stack limit of BPF programs.

Right, that's what I realized and that's why I propose two better alternative. BPF ringbuf is the best variant, but it sets higher minimum kernel requirements.

Reply:
I've read your blog about BPF ringbuf, that's awesome. Thanks.

Since the BPF ringbuf requires bump kernel version and we have a bunch of tools using perfbuf already, let's
use per-cpu array here. WDYT ?